### ISSUE TITLE
External Texture Support for Vulkan

### ISSUE THREAD

**Issue by anlumo** [flutter/flutter#117937]
## Use case

I want to use Flutter as the UI layer in a desktop application that relies heavily on graphics, similar to Adobe Illustrator.

Displaying my documents is planned to happen using Vulkan and Metal in order to allow more flexibility with GPU-based effects (think Photoshop filters). The code for this is native (written in Rust using [wgpu](https://wgpu.rs/)). \[Ideally, I'd like to use DX12 on Windows, but that's not an option with Flutter at all at the moment.]

In order to achieve this, I want to use the embedder API in Vulkan mode. I've written a proof-of-concept [here](https://github.com/anlumo/flutter_embedder), which already works.

However, in order to display the document view, I have to use Flutter's platform view API and integrate with that. This would work fine in theory, however this means that I have to render the view synchronously within the `present_layers_callback` callback. This is not really an option for me, because I want to use the rendering engine [bevy](https://bevyengine.org/) (built on top of wgpu) that has its own render loop. Vulkan can support this use case as long as I'm rendering to different surfaces, but within `present_layers_callback` this is not the case. So, I'd have to either render the platform view to a texture or copy all of Flutter's layers into my own textures, so the renderer can composite them at its own pace.

I think that using Flutter as a UI layer for 3D content would also be interesting for games and other media applications (like Google Earth or a CAD tool for example). There are a few solutions for this (like [Neosis GUI](https://www.noesisengine.com/) and [Coherent Gameface](https://coherent-labs.com/products/coherent-gameface/)), but nothing that comes close to Flutter's feature set and performance combination.

## Proposal

For OpenGL and Metal, the Flutter embedder supports external textures. However, this is not implemented for Vulkan. With this support, I could skip a lot of the overhead described above, making the code much simpler and also faster (because Flutter could directly render the texture using the Texture widget).

**Comment by eladmaimoni** [comment]
Hi @anlumo!

I found your post by searching how can flutter interop with a C++ Vulkan rendering engine.
If I get you right, there is something called _embedder API_ and _platform view API_. Can you point me to the relevant resources?

My goal is basically to embed some custom rendering (such as a Vulkan texture) inside a flutter based UI.
What is the process to achieve this? Any details about the process and limitations would be appreciated.

**Comment by anlumo** [comment]
@eladmaimoni [This is the header file that explains everything](https://github.com/flutter/engine/blob/main/shell/platform/embedder/embedder.h). Sadly, it's the only official documentation, and it's very lacking in details (especially the keyboard handling part).

> My goal is basically to embed some custom rendering (such as a Vulkan texture) inside a flutter based UI.
>
> What is the process to achieve this?

That's basically what my code is doing (it's linked in the ticket description), although I'm using a slightly simplified API (WebGPU-based instead of Vulkan) and it's in Rust.

The thing is that it's currently not possible to directly provide a VkTexture to Flutter and let it handle the rest (that's what this ticket is about), but you have to do it manually.

There are two ways to implement rendering with the embedder API: `FlutterVulkanPresentCallback` and supplying a compositor. For platform views, you have to do the latter (which is much more complicated). Here is the official documentation that explains what you have to do:

https://github.com/flutter/engine/blob/2075cf3f86871a249ef2c7cd99d7f86c899802e9/shell/platform/embedder/embedder.h#L1850-L1864

So, you don't supply a texture, but you get a list of textures containing Flutter stuff to render yourself, and then you're free to display the platform views in whatever way you want.

Note that the documentation doesn't mention that the Flutter Engine expects the backing store texture you supply to it to be cleared to **white** upon initialization, and that it expects to reuse the render buffer from last frame for a new frame (it only sends things that have changed, leaving the rest of the texture fully transparent).

**Comment by eladmaimoni** [comment]
@anlumo THANK YOU!
I hope this issues gets priority.

**Comment by MindStudioOfficial** [comment]
Hi, I am also facing a similar problem with my application.

My goal is to render something using Vulkan and display it as part of my Flutter UI using a Texture Widget.
Seeing this issue gave me hope that there might be a way to do that.

@anlumo while I don't see any progress here, were you able to achieve your goal with some workaround?

**Comment by jwinarske** [comment]
@anlumo I'm starting the implementation on this in a week or so:
https://github.com/flutter/flutter/issues/144613

**Comment by anlumo** [comment]
> @anlumo while I don't see any progress here, were you able to achieve your goal with some workaround?

While I haven't started with the project where I'm going to need this yet, but my planned workaround is to use PlatformViews instead, as mentioned in the description.

**Comment by MindStudioOfficial** [comment]
> @anlumo I'm starting the implementation on this in a week or so:
> https://github.com/flutter/flutter/issues/144613

Which platforms will this apply to?

**Comment by jwinarske** [comment]
> > @anlumo I'm starting the implementation on this in a week or so:
> > #144613
> 
> Which platforms will this apply to?

Anything that uses the Vulkan embedder API would be able to use it.  Platform agnostic.

**Comment by loic-sharma** [comment]
Samsung's Flutter Tizen has forked the engine and added support for Vulkan external textures through the embedder API. This might be useful prior art:

* https://github.com/flutter-tizen/flutter/blob/23d42e47a3569fd042d44612cfbed7f74abcccbc/engine/src/flutter/shell/platform/embedder/embedder.h#L957-L985
* https://github.com/flutter-tizen/flutter/blob/23d42e47a3569fd042d44612cfbed7f74abcccbc/engine/src/flutter/shell/platform/embedder/embedder.cc#L2349-L2369
* https://github.com/flutter-tizen/flutter/blob/23d42e47a3569fd042d44612cfbed7f74abcccbc/engine/src/flutter/shell/platform/embedder/embedder_external_texture_vulkan.h
* https://github.com/flutter-tizen/flutter/blob/23d42e47a3569fd042d44612cfbed7f74abcccbc/engine/src/flutter/shell/platform/embedder/embedder_external_texture_vulkan.cc 
* https://github.com/flutter-tizen/flutter/blob/23d42e47a3569fd042d44612cfbed7f74abcccbc/engine/src/flutter/shell/platform/embedder/embedder_external_texture_source_vulkan.h
* https://github.com/flutter-tizen/flutter/blob/23d42e47a3569fd042d44612cfbed7f74abcccbc/engine/src/flutter/shell/platform/embedder/embedder_external_texture_source_vulkan.cc

**Comment by xiaowei-guan** [comment]
Yes, flutter-tizen has supported render external texture for vulkan. I can create a PR for this part.
